### ISSUE TITLE
A thread timer expired and brain split on AWS

### ISSUE THREAD

**Issue by ThomasSevestre** [acassen/keepalived#2238]
I'm using keepalived (version 2.2.7) on AWS (Ubuntu 20.04)

I've setup the highest possible priority for keepalived threads :

```
global_defs {
  enable_script_security
  max_auto_priority

  vrrp_no_swap
  checker_no_swap
  vrrp_rt_priority 99
  checker_rt_priority 99
  vrrp_rlimit_rttime 100000
  checker_rlimit_rttime 100000
}
```

I still experience occasional thread timer expirations.

Server A :
```
A thread timer expired 2.783285 seconds ago
(VI_1) Received advert from x.x.x.x with lower priority 100, ours 110, forcing new election
```

Server B :
```
(VI_1) Receive advertisement timeout
(VI_1) Entering MASTER STATE
(VI_1) Master received advert from y.y.y.y with higher priority 110, ours 100
(VI_1) Entering BACKUP STATE
```

The server was idle when the issue occured.
Is there anything more I can do to prevent thread expiration ?

**Comment by pqarmitage** [comment]
I am guessing here (I have never used AWS).

You have set the vrrp process to have as high a priority as possible within your AWS VM, and the VM was idle, therefore keepalived will not be able to be blocked by any other process in the VM. The only scenario that I can see that could cause what you are experiencing is that your VM is not being scheduled to run by the host OS, presumably due to CPU/memory demands of other VMs running on the same physical machine. Is this a possibility?

**Comment by ThomasSevestre** [comment]
Thanks for the confirmation, this config helps but does not solve the issue completely. I believe it is always possible the AWS does not schedule the VM. I believe I should have high CPU steal but I'm not 100% sure. It might be related to the instance type (t* vs c* for instance)

In order to limit the impact of the issue, I'm trying to acheive "noprempt at runtime". Instead of above logs, I would like to have :

Server A:
```
A thread timer expired 2.783285 seconds ago
(VI_1) Received advert from x.x.x.x with lower priority 100, ours 110
(VI_1) Entering BACKUP STATE
```

Server B :
```
(VI_1) Receive advertisement timeout
(VI_1) Entering MASTER STATE
```

I've not found a solution yet, do you known if there a parameter for this ?

**Comment by ThomasSevestre** [comment]
I found a dirty solution with `noprempt` and triggering a restart of keepalived on `master_rx_lower_pri` event...
I would be nice to have a clean solution

**Comment by pqarmitage** [comment]
I need to have a think about this, so that I don't implement something that then causes other problems. What I don't like about this conceptually is that we have a higher priority instance that is (or at least thinks it is) master, and we want it to give way to a lower priority instance this is also in master state; that just feels "wrong" to me.

How about this as an idea:
If there is a `A thread timer expired n.nnnnnn seconds ago` message, for any VRRP instance that is in master state and for which it is now more than three advert intervals since it last sent an advert, the VRRP instance transitions to backup state, since another system may have taken over as master. This feature would only be triggered if `nopreempt` is configured, since otherwise the VRRP instance will take over as master again anyway. I think this would work for you.

The rationale here is that a backup instance will become master _**3 * advert interval + (256 - priority)/256**_ (for VRRPv2) or _**3 * advert interval + (256 - priority) / 256 * advert interval**_ (for VRRPv3) after it last received an advert. So if we haven't sent an advert for 3 advert intervals, then another instance could have become master.

We could improve this by adding a parameter `highest_other_priority` which is the highest priority of any of the other peer VRRP instances. Then after a "thread timer expired a long time ago" event, a master instance would only revert to backup if _3 * advert interval + (256 - highest_other_priority)/256_ seconds (or equivalent for VRRPv3) had elapsed since it last sent an advert. This would mean that the server that experienced the thread timer expired event should only revert to backup if another instance has already taken over as master.

@ThomasSevestre Your thoughts and comments on this, as to whether you think this is sensible and would it work for you, would be appreciated. If you are happy with this, then I will look at implementing it.

P.S. I actually quite like this as a solution, despite my initial comment above about it feeling "wrong".

**Comment by ThomasSevestre** [comment]
If I understand well, you propose to implement the following statement : "If a timer expiration has caused the other instance to because a master, then you force a backup transition".

It should solve the above use case but they are more use cases. A backup node may transition to master because of a thread timout or during a network issue between backup and master nodes (there might be more use cases I'm not aware of).

I have hosts handling long running TCP connections established by natted clients. In this case, every IP failover causes trouble. And in my opinion, every useless transition should be considered an issue. To limit IP failovers, whenever a backup node has transition to master, it should stay the master. Other nodes should not preempt and should transition to backup if they are in master state. 

I believe it is possible to achieve this with the following configuration / behavior :
- state BACKUP
- nopreempt
- lower_prio_no_advert
- transition to backup on master_rx_lower_pri event

3 first points are already possible. Only the last one is missing.

Since keepalived already support nopreempt at startup, it should be safe to transition to backup on master_rx_lower_pri event.

**Comment by pqarmitage** [comment]
> A backup node may transition to master because of a thread timout or during a network issue between backup and master nodes (there might be more use cases I'm not aware of).

A thread timeout or a network issue between master and backups is an _environment failure_ that VRRP is not designed to handle. VRRP is designed for routers, and there is the expectation that they will have sufficient processing and scheduling capability that they can run the protocols configured on them. Attempting to run a time based protocol (as VRRP must be in order to detect the disappearance of a router) in an environment where processes are not scheduled in a timely manner is never going to work reliably. The only answer to that is to ensure that the systems which are running VRRP have sufficient resources to be able to take the necessary actions in a timely manner. A workaround for this _might_ be to increase the advert interval to be long enough that any thread timeouts do not exceed 2 advert intervals.

A network failure is a lower layer failure as far as keepalived/VRRP are concerned, and in order to survive such failures there needs to be resilience at layer 2, such as rapid spanning-tree. VRRP is designed to provide reliability at layer 3, but in order to do that it must be provided with a reliable layer 2 environment, and a reliable processing environment.

> I have hosts handling long running TCP connections established by natted clients. In this case, every IP failover causes trouble. 

_conntrackd_ is probably what you need for this. You would need to run it on all systems on which keepalived is running. If you run conntrackd then failovers should not cause you a problem, since your problem relates to a loss of state information which conntrackd updates between the systems.

> And in my opinion, every useless transition should be considered an issue.

I agree with this statement; however it is the cause of the issue that needs to be resolved, not the symptoms of the issue.

> To limit IP failovers, whenever a backup node has transition to master, it should stay the master. Other nodes should not preempt and should transition to backup if they are in master state.

What you are suggesting here is in violation of the VRRP protocol. If two systems (routers) are both in master state, the VRRP protocol says the the higher priority one should remain master and the lower priority one reverts to backup, whereas you are asking for the lower priority to remain master (which sort of makes it the higher priority to be master, but configured with a lower priority).

Assuming you have nopreempt configured, consider the situation where a lower priority VRRP instance is master, and the higher priority instance is in backup state due to nopreempt. If there is a thread timeout or network issue, then the higher priority instance will transition to master, and then when communication between the VRRP instances starts again, your proposal would cause the more recent master (i.e. the higher priority instance) to revert to backup; this seems to be exactly the situation you are seeking to avoid.

What you really want is for the VRRP instance that most recently transitioned to master to remain in master state and any other instances to revert to backup state. The VRRP protocol doesn't handle this, although it would be possible to create a new protocol, as an enhancement to VRRP, which would allow that. Some mechanism whereby the various systems running keepalived communicate between each other regarding the time of the last master transition per VRRP instance, and modifying the priority of the VRRP instances via track_scripts or better, track_files, could, I think, be made to work.

> I believe it is possible to achieve this with the following configuration / behavior :
> 
>     state BACKUP
>     nopreempt
>     lower_prio_no_advert
>     transition to backup on master_rx_lower_pri event

As described above, point 4 above does not help if a lower priority instance was master before the problem occurred which caused a higher priority instance also to become master.
 
> 3 first points are already possible. Only the last one is missing.
> 
> Since keepalived already support nopreempt at startup, it should be safe to transition to backup on master_rx_lower_pri event. I would say it is not worse that existing nopreempt parameter

nopreempt is part of the VRRP specification, and is designed to minimise master/backup transitions based on the infrastructure over which VRRP is operating being reliable, i.e. VRRP needs a reliable processing and lower network layer environment.

In summary I do not see what more keepalived can do other than what I suggested in my previous post. I think using conntrackd could help stop/minimise the problem of longstanding NAT'd TCP connections being lost, using something like spanning-tree to create layer-2 reliability, and ensuring that keepalived is reliably scheduled within the required times (or by increasing the advert interval as a work around) are the most likely options to solve your problems.

**Comment by ThomasSevestre** [comment]
Thank you very much for your helpfull answer (as always).

I've tested this :

> Assuming you have nopreempt configured, consider the situation where a lower priority VRRP instance is master, and the higher priority instance is in backup state due to nopreempt. If there is a thread timeout or network issue, then the higher priority instance will transition to master, and then when communication between the VRRP instances starts again, your proposal would cause the more recent master (i.e. the higher priority instance) to revert to backup; this seems to be exactly the situation you are seeking to avoid.

You are right, depending on which keepalived instance sends the first advert after the network failure, there might be an unwanted transition or not.

> What you really want is for the VRRP instance that most recently transitioned to master to remain in master state and any other instances to revert to backup state. The VRRP protocol doesn't handle this, although it would be possible to create a new protocol, as an enhancement to VRRP, which would allow that. Some mechanism whereby the various systems running keepalived communicate between each other regarding the time of the last master transition per VRRP instance, and modifying the priority of the VRRP instances via track_scripts or better, track_files, could, I think, be made to work.

You are right again, on second though I can't find easy solution to ensure minimum transition without tracking the time of last master transition. Thus the solution you proposed [here](https://github.com/acassen/keepalived/issues/2238#issuecomment-1380738672) looks good

I'll also use existing master_rx_lower_pri event to take back ownership of the VIP

**Comment by pqarmitage** [comment]
I'll have a look at implementing my suggestion at https://github.com/acassen/keepalived/issues/2238#issuecomment-1380738672.

I would be very interested to know if using conntrackd helps with the issue of open TCP connections continuing to work following a VRRP master change, since it could be useful in other instances as well. It might even be that it could be worth having options in keepalived to manage conntrackd.

**Comment by ThomasSevestre** [comment]
Sorry, conntrackd does not fit with my usage.
It looks like a good candidate to implement TCP HA on the natting device.

**Comment by nser77** [comment]
> I'm using keepalived (version 2.2.7) on AWS (Ubuntu 20.04)
> 
> I've setup the highest possible priority for keepalived threads :
> 
> ```
> global_defs {
>   enable_script_security
>   max_auto_priority
> 
>   vrrp_no_swap
>   checker_no_swap
>   vrrp_rt_priority 99
>   checker_rt_priority 99
>   vrrp_rlimit_rttime 100000
>   checker_rlimit_rttime 100000
> }
> ```
> 
> I still experience occasional thread timer expirations.
> 
> Server A :
> 
> ```
> A thread timer expired 2.783285 seconds ago
> (VI_1) Received advert from x.x.x.x with lower priority 100, ours 110, forcing new election
> ```
> 
> Server B :
> 
> ```
> (VI_1) Receive advertisement timeout
> (VI_1) Entering MASTER STATE
> (VI_1) Master received advert from y.y.y.y with higher priority 110, ours 100
> (VI_1) Entering BACKUP STATE
> ```
> 
> The server was idle when the issue occured. Is there anything more I can do to prevent thread expiration ?

Hi, what's the ```advert_int``` value of your ```vrrp_instance``` configuration?

If set to 1, in a virtualized scenario, you might want to increase the interval due to a possible scheduling issue, as you rightly said; increasing this value to 3 solved our case.

**Comment by pqarmitage** [comment]
Commit 5b6ee8f adds keyword `thread_timer_expired` to the `vrrp_instance` block. If it is specified, and nopreempt is also configured, then if the delay in scheduling the keepalived vrrp process is sufficiently long that another system may have taken over as master, then the delay master instance will revert to backup.

The is one configurable option with `thread_timer_expired` and that is the highest priority of the corresponding instance on any other system (the time delay of another system taking over varies with the priority). So if this system as priority 220, and other systems have priority 200 and 240, then
    `thread_timer_expired 240`
should be configured. The system with priority 240 should have
    `thread_timer_expired 220`
configured.

**Comment by anilech** [comment]
Hi,

> Commit https://github.com/acassen/keepalived/commit/5b6ee8f19a8d15f56a88016a5e61369da1357917 adds keyword `thread_timer_expired`

I've checked the [release notes](https://www.keepalived.org/release-notes/Release-2.2.8.html), they mention this keyword as well.

my keepalived 2.2.8 complains on it:

```
Sep 19 11:04:50 host.domain Keepalived[3582923]: Starting Keepalived v2.2.8 (04/04,2023), git commit v2.2.7-154-g292b299e+
. . .
Sep 19 11:04:50 host.domain Keepalived_vrrp[3582924]: (/etc/keepalived/keepalived.conf: Line 19) Unknown keyword 'thread_timer_expired'
```

The commit itself shows that the different keyword has been used:
```
   2175 +       install_keyword("timer_expired_backup", &vrrp_timer_expired_backup_handler);
```

After replacing the `thread_timer_expired` with the `timer_expired_backup`, keepalived starts without complaining.

Could you please verify which option is correct and if documentation should be updated as well.

BR Alexey.

**Comment by pqarmitage** [comment]
As you state, the commit message and release notes don't match the documentation. Commit 9c7d00c adds the keyword `thread_timer_expired` to the code so that it matches the commit message and release notes, but it logs a message to change the keyword to `timer_expired_backup`. 

The correct keyword to use is `timer_expired_backup`. Apologies for any confusion caused.

BTW if you are experiencing occurrences of thread_timer_expired, then that indicates that keepalived is not being scheduled quickly enough by the kernel, possibly due to the system being too busy.
